api/envoy/extensions/filters/http/grpc_field_extraction/v3/config.proto

```diff
@@ -0,0 +1,143 @@
+//
+// # Config Requirements
+// 1. the target field should be singular primitive type or repeated primitive
+// type and its value will be extracted in string format.
+// 2. the intermediate type could also be repeated.
+//
+// # Output Format
+// The result format will be `field<StringValue>` -> `values<ListValue of StringValue>` in the dynamic metadata<google.protobuf.Struct>.
+//
+// # Performance
+// This filter should be performant as it
+// 1. converts between the gRPC message from EnvoyBuffer without data copy.
+// 2. parse the gRPC message binary directly without deserialization.
+//
+// # Example,
+// we have the following request definition for the gRPC method `/pkg.svc/Method`.
+//
+// message MethodRequest {
+//   string foo = 1;
+//   Nested nested = 2;
+//   ...
+// }
+//
+// message Nested {
+//  repeated string bar = 1;
+// }
+//
+// This is the filter config(expressed in JSON format).
+// {
+//   "descriptor_set":{...},
+//   "extractions_by_method": {
+//      "/pkg.svc/Method":{
+//         "request_field_extractions":{
+//            "foo":{
+//            },
+//            "nested.bar":{
+//            }
+//         }
+//      }
+//   },
+//   ...
+// }
+//
+//
+// During runtime, the filter receives the following `MethodRequest` message(expressed in JSON format).
+// {
+//   foo: "val_foo",
+//   nested: { "bar": ["val_bar1", "val_bar2"]}
+// }
+//
+// The filter will writes the following dynamic metadata(expressed in JSON format).
+//
+// `envoy.filters.http.grpc_field_extraction`: {
+//   "foo":[
+//      "val_foo"
+//   ],
+//   "nested.bar":[
+//      "val_bar1", "val_bar2"
+//   ]
+// }
+
+message GrpcFieldExtractionConfig {
+  // The proto descriptor set binary for the gRPC services.
+  //
+  // It could be passed by a local file through `Datasource.filename` or embedded in the
+  // `Datasource.inline_bytes`.
+  config.core.v3.DataSource descriptor_set = 1 [(validate.rules).message = {required: true}];
+
+  // Specify the extraction info.
+  // The key is the fully qualified gRPC method name.
+  // `/$package.$Service/$Method`, like
+  // `/endpoints.examples.bookstore.BookStore/GetShelf`
+  //
+  // The value is the field extractions for individual gRPC method.
+  map<string, FieldExtractions> extractions_by_method = 2;
+}
+
+// This message can be used to support per route config approach later even
+// though the Istio doesn't support that so far.
+message FieldExtractions {
+  // The field extractions for requests.
+  // The key is the field path within the grpc request.
+  // For example, we can define `foo.bar.name` if we want to extract
+  // Request.foo.bar.name.
+  // message Request {
+  //   // The namespace in which the Workspace should be created.
+  //   Foo foo = 1;
+  // }
+  //
+  // message Foo {
+  //   Bar bar = 1;
+  // }
+  // message Bar {
+  //   string name = 1;
+  // }
+  map<string, RequestFieldValueDisposition> request_field_extractions = 1;
+}
+
+message RequestFieldValueDisposition {
```

Comment:
Is this actually needed? When would someone want to change metadata name?

Reply:
If the proxy serves multiple PDP requiring extracting multiple fields for one request, we want to decouple the disposition.